DMD.h

```diff
@@ -0,0 +1,177 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2021, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+// Description:Computes the DMD algorithm on the given snapshot matrix.
+
+#ifndef included_DMD_h
+#define included_DMD_h
+
+#include <vector>
+#include <complex>
+
+namespace CAROM {
+
+class Matrix;
+class Vector;
+
+class DMD
+{
+public:
+
+    /**
+     * @brief Constructor.
+     */
+    DMD(int dim);
+
+    /**
+     * @brief Sample the new state, u_in.
+     *
+     * @pre u_in != 0
+     *
+     * @param[in] u_in The state at the specified time.
+     *
+     * @return True if the sampling was successful.
+     */
+    bool takeSample(double* u_in);
+
+    /**
+     * @param[in] energy_fraction The energy fraction to keep after doing SVD.
+     */
+    void train(double energy_fraction);
+
+    /**
+     * @param[in] k The number of modes (eigenvalues) to keep after doing SVD.
+     */
+    void train(int k);
+
+    /**
+     * @brief Predict new initial condition using d_phi.
+     *
+     * @param[in] init The initial condition.
+     */
+    void projectInitialCondition(const Vector* init);
+
+    /**
+     * @brief Predict state given a time. Uses the projected initial condition of the
+     *        training dataset (the first column).
+     *
+     * @param[in] t The time of the outputted state.
+     * @param[in] dt The delta time of the simulation.
+     */
+    Vector* predict(double t, double dt);
```

Comment:
My understanding is that `n = t/dt` is used, not `t` or `dt` individually. Would it make more sense to have one input for `n` (which could be a `double`), representing the number of online timesteps? Isn't `dt` associated with an offline simulation and assumed to be the same for online? Variable dt is not allowed, right? Or could variable dt work somehow?

Reply:
You're right. I think this would work better.